Pass ar5 tests

owenr opened the pull request:
Update test code to address deprecations/warnings and fix one legitimate issue with AR5. Not happy with several aspects of this change:
* Gem `rails-controller-testing` does not work as advertised (possibly an issue with load order?) and I could only get it working when manually included, see `test_helper.rb`
* HTTP request test methods in Rails 5 need parameter hash keys named 'params', so a new wrapper function is introduced twice (once in test_helper, again in spec_helper)
* Addressing the AR5 issue seems to require a way of preventing #save on the has_one object, and there is almost certainly a better way than temporarily redefining #save on the instance

Having said that, the changes pass tests locally on MRI 2.3 (ar5) and 1.9.3-p551 (ar3/ar4). Comments and suggestions welcome.

@jaredbeck

owenr commented:
Having some issues with tests intermittently failing. Any ideas on that are also welcome.

bronson commented:
Interesting...  Ten tests failed the first time I ran `rake`, but that was it.  Now I can't get any to fail.

Here's the log: https://gist.github.com/bronson/70dd3c4aaf395df4bf93

And, if I `rm -r test/dummy/db/*.sqlite3` and run `rake` again, I get the same ten errors.  Looks to be an issue related to a fresh database?

Is that the problem you're seeing?

bronson commented:
I'm at 25 iterations, no failures yet.  I'll let it run for a few hundred.
```bash
ctr=; while rake; do echo ITERATION $((ctr += 1)); done
```
But, no, I'm not seeing any intermittent errors on your branch.

I'm using sqlite, are you testing on a different database?

jaredbeck commented:
Owen, thanks for this! I'll review what I can this week.

>  .. if I rm -r test/dummy/db/*.sqlite3 and run rake again, I get the same ten errors. Looks to be an issue related to a fresh database?

Scott, Yeah, that's a known issue with running the local test suite. If you can fix it, that'd be great.

bronson commented:
260 iterations, 0 failures.  I don't see any intermittent issues.  (sqlite, mac, ruby 2.2.2)

owenr commented:
Initially tested using sqlite3 locally, no issues. As you can see, Travis CI failed all my builds for various reasons; it looks like Postgresql is the main issue now. Simulating locally it took me 22 runs before this failure (postgres, mac, 2.3.0):

```
PaperTrailCleanerTest#test_: `clean_versions!` method No options provided should removes the earliest version(s).  [/Users/owen/dev/paper_trail/test/unit/cleaner_test.rb:34]:
--- expected
+++ actual
@@ -1 +1 @@
-["Aglae Schaden", "Carley Doyle", "Karianne McDermott"]
+["Aglae Schaden", "Carley Doyle", "Jaime Witting"]
```

bronson commented:
Switched to postgres and I finally got a failure after 76 iterations: https://gist.github.com/bronson/01f05504f81efe072b40

Glancing at it, maybe it died because name contained a single quote... A sql injection opportunity?  :)  I'll look closer tomorrow if nobody beats me to it.

Also, there might be a test order dependency.  At least:

`rspec --seed=0`: 0 errors
`rspec --seed=1`: 1 error
`rspec --seed=9`: 10 errors

Do the tests have a required order?

owenr commented:
Confirmed, unescaped input (only when using postgres). I've written some failing tests and will open an issue. Unfortunately it seems unrelated to the failures from Travis [here](https://travis-ci.org/airblade/paper_trail/jobs/102019148#L355) and [here](https://travis-ci.org/airblade/paper_trail/jobs/101750643#L355).

I'm not familiar with the impact of seeding on this test suite and I couldn't reproduce those spec failures with seeds 1 and 10 (mac, postgres, ruby 2.2.2).

jaredbeck commented:
When comparing gem versions, I like to use `Gem::Version`, e.g.

```diff
- ActiveRecord::VERSION::STRING >= '5.'
+ Gem::Version.new(ActiveRecord::VERSION::STRING) >= Gem::Version.new("5")
```

PS: If we didn't have to support AR3, we could use the more convenient `ActiveRecord.gem_version`

jaredbeck commented on lib/paper_trail/reifier.rb:
```diff
@@ -162,6 +164,15 @@ def reify_has_ones(transaction_id, model, options = {})
         end
       end
 
+      # Suppress #save so we can reassociate with the reified master of a has_one relationship
```
Can you expand on this comment? Specifically, please explain what changed in AR5.

Also, please alphabetize methods.

jaredbeck commented:
Owen, I've been cherry-picking the less controversial commits out of here.

- https://github.com/airblade/paper_trail/pull/691
- https://github.com/airblade/paper_trail/pull/692

I thought there was no sense in waiting, but it will mean you'll have to rebase this PR, sorry.

owenr commented:
No problem at all; like I said, this code is not right and should be improved before merging.

Some comments:
* I've reordered methods alphabetically as requested.
* Switched to Gem::Version for comparisons, but note that comparison with '5' is insufficient ('5' is not >= '5.0.0.beta').
* I can't find the AR5 change to has_one, but from the symptoms it looks like the associated object is persisted before the main object (which seems correct as per the AR4 documentation). I've elaborated on the comment.

jaredbeck commented:
Owen, It looks like this needs another rebase, sorry.

jaredbeck commented on test/test_helper.rb:
```diff
@@ -16,6 +16,17 @@ def using_mysql?
+    ActionDispatch::IntegrationTest.send(:include, Rails::Controller::Testing::Integration)
+
+    ActionView::TestCase.send(:include, Rails::Controller::Testing::TemplateAssertions)
+  end
```
This whole `on_load` could use a comment, e.g. why are we including these modules, why are we using `send`, why `on_load`, etc. I don't personally understand any of it, so the comment is mostly for me :p

owenr replied:
Copied wholesale from [rails-controller-testing](https://github.com/rails/rails-controller-testing/blob/master/lib/rails-controller-testing.rb). I don't know why putting it here works but requiring the library doesn't. I've also tried `require: false` in the Gemfile and manually requiring it instead but that doesn't help. Removing this code, I get:
```
NoMethodError: assigns has been extracted to a gem. To continue using it,
        add `gem 'rails-controller-testing'` to your Gemfile.
```
Not knowing if this is caused by something in PaperTrail or an issue with rails-controller-testing itself (or Rails 5), I'm not sure whether I should raise an issue over there.

jaredbeck commented on test/test_helper.rb:
```diff
@@ -89,6 +100,19 @@ def change_schema
+  else
+    args
+  end
+end
```
Please add a comment explaining why this `params_wrapper` method is necessary.

jaredbeck commented:
> HTTP request test methods in Rails 5 need parameter hash keys named 'params'

Is this change documented in a changelog somewhere, or a guide?

jaredbeck commented:
>> HTTP request test methods in Rails 5 need parameter hash keys named 'params'
> 
> Is this change documented in a changelog somewhere, or a guide?

Found it. It's in the release notes for ActionPack 5:

> Migrating to keyword arguments syntax in ActionController::TestCase and ActionDispatch::Integration HTTP request methods. (Pull Request)
> http://edgeguides.rubyonrails.org/5_0_release_notes.html#action-pack

jaredbeck commented on spec/spec_helper.rb:
```diff
@@ -102,3 +102,11 @@
+  else
+    args
+  end
+end
```
Please add a comment explaining the change to ".. keyword arguments syntax in ActionController::TestCase and ActionDispatch::Integration HTTP request methods." and include a link to the ActionPack changelog (https://github.com/tgxworld/rails/blob/master/actionpack/CHANGELOG.md)

owenr commented:
Rebased. I also fixed one more deprecation warning (use_transactional_tests) and added a note about 1-second differences in tests (for MySQL and AR5). Still not happy.

owenr commented:
I think this is just about done. Please review and let me know; I'll squash the commits.

jaredbeck commented:
This looks good to me. I will merge this unless Ben has any concerns @batter 

I wish we didn't have to time-travel just to make MySQL happy, but it is something we've run into before (https://github.com/airblade/paper_trail/issues/314, https://github.com/airblade/paper_trail/commit/1d120ab9e4cb02890970b69e8a4e6e994b5fd117)

owenr commented:
Time travel is required due to a change in ActiveRecord 5 (if no attributes change, nothing is done). We could `sleep` instead of travel if that's preferred.
I've squashed the commits (and backed out an unneeded change to rubocop_todo picked up in review).

jaredbeck commented:
Merged. Thanks for your work on this, Owen.

bronson commented:
Wooo!  Great job Owen.
